core/src/main/java/com/linecorp/armeria/internal/common/stream/SurroundingPublisher.java

```diff
@@ -309,6 +309,8 @@ private void publishDownstream(T item, boolean head) {
                 if (requested < Long.MAX_VALUE) {
                     requested--;
                 }
+                subscribed = true;
+                publisher.subscribe(this, executor, options);
```

Comment:
✅ updated, subscribe upstream preemptive on sendHead().

If not, below scenario failed
```
// elements: 1, head -> empty() -> end (no tail)
1. request(1)
2. sendHead() // requested: 1 -> 0
3. publish() -> requested: 0 -> return; // expectComplete(), but no onComplete()
```